ci/Jenkinsfile

```diff
@@ -0,0 +1,116 @@
+@Library('gpuci_shared_lib') _
+
+pipeline {
+  agent any
+  options {
+    buildDiscarder(logRotator(numToKeepStr: '10'))
+    ansiColor('xterm')
+  }
+  stages {
+    stage("Style Checks") {
+      steps {
+        sh "env"
+      }
+    }
+    stage("C++ Matrix Build") {
+      environment {
+        BUILD_MODE = 'pull-request'
+        PARALLEL_LEVEL = '4'
+        CUDA="11.5"
+      }
+      steps {
+        script {
+          parallel generateStage("cuda_build", {
+            checkout scm
+
+            withCredentials([[
+                $class: 'AmazonWebServicesCredentialsBinding',
+                credentialsId: "aws-s3-gpuci",
+                accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+            ]]) {
+              sh "./ci/build_cpp.sh"
+            }
+          })
+        }
+      }
+    }
+
+    stage("Python Matrix Build") {
+      environment {
+        CUDA="11.5"
+      }
+      steps {
+        script {
+          parallel generateStage("python_build", {
+            cleanWs (
+              deleteDirs: true,
+              externalDelete: 'sudo rm -rf %s'
+            )
+            checkout scm
+
+            withCredentials([[
+                $class: 'AmazonWebServicesCredentialsBinding',
+                credentialsId: "aws-s3-gpuci",
+                accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+            ]]) {
+              sh "./ci/build_py.sh"
+            }
+          })
+        }
+      }
+    }
+
+    stage("PR/Branch Test Stage") {
+      when { not { triggeredBy 'TimerTrigger' } }
+      environment {
+        CUDA="11.5"
+        UPLOAD_PKGS = """${sh(
+              returnStdout: true,
+              script: 'echo "$GIT_BRANCH" | grep -iq "pr" && echo "false" || echo "true"'
+          )}"""
```

Comment:
In hindsight, I think `UPLOAD_PKGS` can be removed. Its value can ultimately be inferred from `BUILD_MODE`.

We'll have to evaluate how these values are being used in the scripts currently.